src/main/scala/is/hail/variant/Genotype.scala

```diff
@@ -177,8 +178,56 @@ class RowGenotype(r: Row) extends Genotype {
 
 object Genotype {
   def buildGenotypeExtractor(t: Type): (Any) => Genotype = {
-    assert (t == TGenotype)
-    (a: Any) => a.asInstanceOf[Genotype]
+    if (t == TGenotype)
+      (a: Any) => a.asInstanceOf[Genotype]
+    else {
+      def queryOpt[T >: Null](field: String)(implicit hr: HailRep[T]): Option[(Any) => T] = {
+        try {
+          val (ft, fq0) = t.queryTyped(field)
+          if (ft == hr.typ)
+            Some((a: Any) => fq0(a).asInstanceOf[T])
+          else
+            None
+        } catch {
+          case _: AnnotationPathException =>
+            None
+        }
+      }
+
+      val gtq = queryOpt[java.lang.Integer]("GT")(callHr)
+      val adq = queryOpt[IndexedSeq[Int]]("AD")
+      val dpq = queryOpt[java.lang.Integer]("DP")
+      val gqq = queryOpt[java.lang.Integer]("GQ")
+      val plq = queryOpt[IndexedSeq[Int]]("PL")
+      val gpq = queryOpt[IndexedSeq[Double]]("GQ")
+
+      if (plq.isDefined || gpq.isEmpty)
+        (a: Any) =>
+          new RowGenotype(Row(
+            gtq.map(_ (a)).orNull,
```

Comment:
Yes.  This is so terrible my eyes are bleeding.  It is also incredibly slow.  As soon as your unsafe stuff goes in I will start integrating it into these towards_...  branches.